L1Trigger/TrackFindingTMTT/interface/L1KalmanComb.h

```diff
@@ -0,0 +1,208 @@
+///=== This is the base class for the Kalman Combinatorial Filter track fit algorithm.
+
+#ifndef __L1_KALMAN_COMB__
+#define __L1_KALMAN_COMB__
+
+#include <TMatrixD.h>
+#include "L1Trigger/TrackFindingTMTT/interface/TrackFitGeneric.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Stub.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Settings.h"
+#include "L1Trigger/TrackFindingTMTT/interface/L1track3D.h"
+#include "L1Trigger/TrackFindingTMTT/interface/L1fittedTrack.h"
+#include "L1Trigger/TrackFindingTMTT/interface/KalmanState.h"
+#include <map>
+#include <vector>
+#include <fstream>
+#include <TString.h>
+
+class TH1F;
+class TH2F;
+
+namespace TMTT {
+
+  class TP;
+  class KalmanState;
+  class StubCluster;
+
+  class L1KalmanComb : public TrackFitGeneric {
+  public:
+    enum OVERLAP_TYPE { TYPE_NORMAL, TYPE_V2, TYPE_NOCLUSTERING, TYPE_TP };
+
+  public:
+    L1KalmanComb(const Settings *settings, const uint nPar, const string &fitterName = "", const uint nMeas = 2);
+
+    virtual ~L1KalmanComb() {
+      this->resetStates();
+      this->deleteStubClusters();
+    }
+
+    L1fittedTrack fit(const L1track3D &l1track3D);
+    void bookHists();
+
+  protected:
+    static std::map<std::string, double> getTrackParams(const L1KalmanComb *p, const KalmanState *state);
+    virtual std::map<std::string, double> getTrackParams(const KalmanState *state) const = 0;
+
+    // Get track params with beam-spot constraint & chi2 (r-phi) after applying it..
+    virtual std::map<std::string, double> getTrackParams_BeamConstr(const KalmanState *state,
```

Comment:
why not ?

Reply:
4.4
http://cms-sw.github.io/cms_coding_rules.html

Reply:
That's rather arbitrary, and doesn't give any reason why not.

There is a lot of work about de-virtualisation of function calls in the compilers (in LLVM at least), which tries to let the compiler infer when it can bypass the v-table and make a direct function call; there are also cases where you use an object by value, rather than by pointer/reference, and virtual functions can be called directly.

That said, `inline` does not mean "inline" any more, it's just a way to say "this function definition will appear in multiple translation units" to the compiler and linker.

So, back to my question, is there a non-arbitrary reason why not ?

Reply:
From this PR it is not clear how this function is going to be called, from an object by value or from an object by pointer/reference.   I hope that the CMS rule _not to inline virtual function_  is precisely to avoid the latter cases. So, I would rather not see this inline definition.

If the argument is that with the new compiler the misuse is automatically detected, and that this rule 4.4 is no more relevant for CMS, we should remove it.

Reply:
That's not the point - calling the function through an object/reference is not a misuse, even if the virtual function is inline.

AFAIK, the only thing that should be avoided is to **declare** a `virtual` function as not `inline` in the class body in a header file, and them **define** it as inline in the source file.
E.g., this should be avoided:
`file.h`
```c++
class Object {
  virtual void doSomething();
};
```
`file.cc`
```c++
#include "file.h"

inline
void Object::doSomething() { ... }
```
According to the GCC documentation
> Make sure that any inline virtuals are declared inline in the class body, even if they are not defined there.

Reply:
What do you conclude? No-where do we actually use the "inline" keyword, so it's up to the compiler to decide what it wants to do. Also note that the virtual functions mentioned in the original comment are pure virtual. I already see hundreds of pure virtual functions like this declared in header files elsewhere in CMSSW.

Reply:
> No-where do we actually use the "inline" keyword, so it's up to the compiler to decide what it wants to do.

Not really: by defining the function within the class body, it is implicitly declared as `inline`. This come from C++, not from the compiler.

> Also note that the virtual functions mentioned in the original comment are pure virtual.
> I already see hundreds of pure virtual functions like this declared in header files elsewhere in CMSSW.

I'm not sure what you mean - a pure virtual function cannot be *defined*, so it doesn't apply to this discussion.

Reply:
> > No-where do we actually use the "inline" keyword, so it's up to the compiler to decide what it wants to do.
> 
> Not really: by defining the function within the class body, it is implicitly declared as `inline`. This come from C++, not from the compiler.

And they keyword `inline` really tends to mean "linker please remove duplicates of this symbol", actual inlining indeed being up to the compiler.